### PR TITLE
Integrate rational PoSt construction

### LINKED ISSUE #3307: Integration Rational PoSt construction
### Description
The rust-fil-proofs team are close to finishing implementation of the Rational PoSt construction. We wish to integrate this as soon as possible so as to include it in the long-running network release scheduled for late September.

For more background, see https://github.com/filecoin-project/rust-fil-proofs/issues/849

This issue is scoped specifically to work that needs to happen before the network is started. Work that can be deferred for a protocol upgrade of the running 0.5 network is less urgent, but we need to prepare now to make such work possible.

### Acceptance criteria
- go-filecoin uses the forthcoming Rational PoSt construction

### Risks + pitfalls
- proofs implementation is not yet complete, yet we want to start integration ASAP
- The rust-fil-proofs FFI may change
- Many PoSt characteristics may change dramatically (e.g. PoSt computation time), requiring protocol parameter changes to accomodate

### PR DESCRIPTION
**Pull request by acruikshank**
closes #3307

### Problem

We need to use the Rational PoSt construction. See https://github.com/filecoin-project/specs/blob/master/proof-of-spacetime.md

### Solution

This PR upgrades `rust-fil-proofs`, `rust-fil-sector-builder`, and `go-sectorbuilder` to the new rational PoSt construction. It also makes the changes required in go-filecoin to support the new method signatures. 

This PR does not implement all the changes the new post construction requires of go-filecoin. Specifically it does not alter the schedule on which PoSts are submitted or storage slashing for submitting outside this schedule. It also does not check for faults. All of this work could be deferred to a protocol upgrade after the release, since the required proofs interfaces are now in place.

### Changes to go-sectorbuilder

```
-----------------------------------------------
commit dc3514247d6620d49a58c01db241bab7d6007e82
Merge: b1ee425 8b7ae87
Author: Erin Swenson-Healey <l@s3r.me>
Date:   Thu Aug 29 09:26:49 2019 -0700

    Merge pull request #16 from filecoin-project/feat/rational-post
    
    updates to consume new Rational PoSt API


-----------------------------------------------
commit 8b7ae87d9688507870ea6a6fb500e5348b8249d4
Author: laser <l@s3r.me>
Date:   Wed Aug 28 14:51:42 2019 -0700

    build(print rust log to stderr when running tests)


-----------------------------------------------
commit cc63aae5d38b29e87e862a31e6cbbcfdb38537e4
Author: laser <l@s3r.me>
Date:   Wed Aug 28 14:23:18 2019 -0700

    bug(wrap in parens to ensure message sent to stderr)


-----------------------------------------------
commit ea4ff72c418eda796393936f866e9ff708b749f6
Author: laser <l@s3r.me>
Date:   Wed Aug 28 14:16:37 2019 -0700

    bug(install paramfetch from HEAD of master if not available in release tarball)


-----------------------------------------------
commit a6545b01954e088d9bbf781afd966ac05e1e1ddc
Author: laser <l@s3r.me>
Date:   Wed Aug 28 13:56:49 2019 -0700

    feat(upgrade rust-fil-sector-builder)


-----------------------------------------------
commit 655646ffe9b0ee6a477ccc4091ef7a13d5601b47
Author: laser <l@s3r.me>
Date:   Wed Aug 28 13:20:24 2019 -0700

    bug(appease linter by deleting dead code)


-----------------------------------------------
commit 7ac74f8157573941bd5df38311572f6b14a90905
Author: laser <l@s3r.me>
Date:   Wed Aug 28 12:16:33 2019 -0700

    feat(enforce sort order-invariant with type system)


-----------------------------------------------
commit 8bec037d6a0195c3280b39ce579ac93420cb5e0f
Author: laser <l@s3r.me>
Date:   Wed Aug 28 12:16:11 2019 -0700

    feat(update to consume head of master instead of branch head)


-----------------------------------------------
commit 5e835a05446d0057ab74e592c62f82174b012768
Author: laser <l@s3r.me>
Date:   Wed Aug 28 11:38:44 2019 -0700

    feat(updates to conform to new API)
    
    - sector id is now a uint64
    - generate_post accepts faults instead of producing them
    - generate_post produces a single proof instead of a concatenation of proofs
    - verify_post accepts a single proof


-----------------------------------------------
commit 8060bb800b04de55a8a9f8dae009b4524e95055e
Author: laser <l@s3r.me>
Date:   Wed Aug 28 10:37:44 2019 -0700

    feat(consume new rational PoSt changeset)


-----------------------------------------------
commit b1ee42550ab59cec6c9bd0f3c22fe82fa34f4760
Merge: 186d3b6 bf036d7
Author: Sidney Keese <sidney@carbonfive.com>
Date:   Wed Aug 28 13:03:55 2019 -0700

    Merge pull request #14 from filecoin-project/features/comprehensive-circle-ci
    
    Features/comprehensive circle ci


-----------------------------------------------
commit bf036d7f7c7e8f30cfad8616f36fd5636fa763d1
Author: Sidney Keese <sidke.z@gmail.com>
Date:   Wed Aug 28 12:56:24 2019 -0700

    don't negative substring


-----------------------------------------------
commit 576cd6860a4cbe63b77c5ee2ded67af056a8da53
Author: Sidney Keese <sidke.z@gmail.com>
Date:   Wed Aug 28 11:17:41 2019 -0700

    stderr echo


-----------------------------------------------
commit b3a5d9e60bb5815a9ae17f5faeabe484f846c248
Author: Sidney Keese <sidke.z@gmail.com>
Date:   Wed Aug 28 10:22:14 2019 -0700

    install paramfetch from rfsb


-----------------------------------------------
commit 80ccd68a16644c610cba817d4ca4357485f1eeaf
Author: Sidney Keese <sidke.z@gmail.com>
Date:   Mon Aug 26 11:21:04 2019 -0700

    bump submodule version


-----------------------------------------------
commit 71f3ec5f1e499a265bcec86a84502bc02a2cb65d
Author: Sidney Keese <sidke.z@gmail.com>
Date:   Mon Aug 19 10:51:02 2019 -0700

    comprehensive circleci configuration
    
    add go/rust path to linux env
    
    fix ineffectual assign to err
    
    change working directory
    
    cargo install filecoin-proofs
    
    save pkg to tmp
    
    install pkg-config
    
    fetch and cache parameters
    
    maybe we don't need arch
    
    install jq
    
    wrap paramfetch with with-dots
    
    maybe spell version correctly
    
    reorder linting
    
    use retry script
    
    reordering linting was a mistake
    
    better with-dots.sh

```

### Changes to rust-fil-sector-builder

```
-----------------------------------------------
commit 0282a636f7911bcfd0348af5494da5932ac5808b
Author: Erin Swenson-Healey <l@s3r.me>
Date:   Wed Aug 28 13:55:58 2019 -0700

    bug(make sure to install paramfetch using head of master branch) (#43)


-----------------------------------------------
commit f9868b2bca5a270b766e15e32781494e4ad74bd7
Author: Friedel Ziegelmayer <dignifiedquire@users.noreply.github.com>
Date:   Wed Aug 28 21:13:38 2019 +0200

    feat: update for rational post (#23)
    
    * feat: updates for rational post
    
    * feat(update to consume new PoSt API + sector id)
    
    * refactor(move symbols around in FFI test to make it easier to read)
    
    * refactor(point crates to head of master)
    
    * refactor(remove unused symbols)
    
    * bug(bump cache key to consume new Groth parameters)
    
    * bug(install filecoin-proofs from head of master, which mimics Cargo dependency)


-----------------------------------------------
commit da6d61f6910828939e5695dd1903b16e312bff31
Author: Erin Swenson-Healey <l@s3r.me>
Date:   Tue Aug 27 13:50:36 2019 -0700

    build(point to master instead of crates.io release) (#42)


-----------------------------------------------
commit 604b487673fa52744218abbe45972c1ee0b8d8cb
Merge: d92eeaf ae0d9d5
Author: Sidney Keese <sidney@carbonfive.com>
Date:   Fri Aug 23 16:04:31 2019 -0700

    Merge pull request #40 from filecoin-project/features/include-filecoin-binaries-12
    
    obtain filecoin proofs bins from cargo install


-----------------------------------------------
commit ae0d9d515276ce5e4bd407373193581c673e5832
Author: Sidney Keese <sidke.z@gmail.com>
Date:   Fri Aug 23 14:29:17 2019 -0700

    obtain filecoin proofs bins from cargo install
    
    and package them in a release

```

### Changes in rust-fil-proofs

```
-----------------------------------------------
commit 037d5fb8f2fad474c97f8fb12ababd35fc44edf3
Author: Friedel Ziegelmayer <dignifiedquire@users.noreply.github.com>
Date:   Wed Aug 28 23:42:04 2019 +0200

    feat(filecoin-proofs): paramfetch: update default ipget to 0.4.0 (#851)


-----------------------------------------------
commit b1ba28e9421f63e06cdb312f838239a309bbb616
Author: Erin Swenson-Healey <l@s3r.me>
Date:   Wed Aug 28 14:23:44 2019 -0700

    bug(classic misdirection) (#852)


-----------------------------------------------
commit cb54e8c4d682cffd8e5d0aff58476bf9746c468b
Author: Friedel Ziegelmayer <dignifiedquire@users.noreply.github.com>
Date:   Wed Aug 28 17:18:00 2019 +0200

    fix(filecoin-proofs): properly retrieve commitments in verify_post


-----------------------------------------------
commit 5ca1740929482a9a4875b1d1cb242f636a26d1d1
Author: Erin Swenson-Healey <l@s3r.me>
Date:   Wed Aug 28 02:45:24 2019 -0700

    fix(filecoin-proofs): use SectorId get_unsealed_range


-----------------------------------------------
commit d2b39cc49640370595c0456091ef87b64cb0cc4e
Author: Friedel Ziegelmayer <dignifiedquire@users.noreply.github.com>
Date:   Tue Aug 27 22:39:30 2019 +0200

    feat: implement and switch to Rational PoSt
    
    BREAKING API CHANGES:
    
    - filecoin-proofs: generate_post
    - filecoin-proofs: verify_post
    
    Closes #497


-----------------------------------------------
commit 7b7bf92bce8b9af74ee96fe2208c30dbaab13be0
Merge: db7c602 534d37e
Author: Friedel Ziegelmayer <dignifiedquire@users.noreply.github.com>
Date:   Tue Aug 27 17:50:39 2019 +0200

    Merge pull request #842 from filecoin-project/hotfix/decrease-post-sectors-count
    
    backport hotfix branch into master


-----------------------------------------------
commit 534d37ec7c798dd7847aadda7d03e0f689840a2a
Merge: 0f1fcb9 b40d12a
Author: Lucas Molas <schomatis@gmail.com>
Date:   Mon Aug 26 18:12:53 2019 -0300

    Merge branch 'master' into hotfix/decrease-post-sectors-count


-----------------------------------------------
commit 0f1fcb9b6e491bdb935f9ab26b4bf541b10d27d2
Author: laser <l@s3r.me>
Date:   Thu Aug 22 17:00:53 2019 -0700

    need release binaries

```

### CODE REVIEW

File: protocol/storage/miner_test.go [hunk 1]
```diff
@@ -467,7 +467,10 @@ func TestOnNewHeaviestTipSet(t *testing.T) {
 
 		// assert proof generated in sector builder is sent to submitPoSt
 		require.Equal(t, 3, len(postParams))
-		assert.Equal(t, []types.PoStProof{[]byte("test proof")}, postParams[0])
+		expectedBytes := []byte("test proof")
```
Comment (ZenGround0) [inline protocol/storage/miner_test.go]: nit: doesn't really matter but I think you could be a bit more succinct here by directly instantiating expected.  If I'm wrong about that you could also just cast in one line ¯\_(ツ)_/¯

File: protocol/storage/prover.go [hunk 2]
```diff
@@ -37,7 +37,7 @@ type ProofReader interface {
 type ProofCalculator interface {
 	// CalculatePoSt computes a proof-of-spacetime for a list of sector ids and matching seeds.
 	// It returns the Snark Proof for the PoSt and a list of sector ids that failed.
-	CalculatePoSt(ctx context.Context, sortedCommRs proofs.SortedCommRs, seed types.PoStChallengeSeed) ([]types.PoStProof, []uint64, error)
+	CalculatePoSt(ctx context.Context, sortedCommRs go_sectorbuilder.SortedSectorInfo, seed types.PoStChallengeSeed) (types.PoStProof, error)
```
Comment (ZenGround0) [inline protocol/storage/prover.go]: nit: here and elsewhere the name `sortedCommRs` could be updated to the new thing for heightened clarity.

File: proofs/sectorbuilder/interface.go [hunk 3]
```diff
@@ -90,12 +89,11 @@ type SealedSectorMetadata struct {
 
 // GeneratePoStRequest represents a request to generate a proof-of-spacetime.
 type GeneratePoStRequest struct {
-	SortedCommRs  proofs.SortedCommRs
-	ChallengeSeed types.PoStChallengeSeed
+	SortedSectorInfo go_sectorbuilder.SortedSectorInfo
+	ChallengeSeed    types.PoStChallengeSeed
 }
 
 // GeneratePoStResponse contains PoST proof and any faults that may have occurred.
```
Comment (anorth) [inline proofs/sectorbuilder/interface.go]: Update this comment, as my understanding is the faults aren't coming back here.

File: tools/iptb-plugins/filecoin/local/localfilecoin.go [hunk 4]
```diff
@@ -47,6 +47,9 @@ const defaultSectorsPath = "sectors"
 // defaultLogLevel is the value that will be used for GO_FILECOIN_LOG_LEVEL
 const defaultLogLevel = "3"
 
+// defaultRustLogLevel is the value that will be used for RUST_LOG
+const defaultRustLogLevel = "info"
```
Comment (anorth) [inline tools/iptb-plugins/filecoin/local/localfilecoin.go]: This changes seems unrelated, and also likely to add lots of noise back to our logs. Was it just for debugging?

File: proofs/sectorbuilder/interface.go (hunk 3 above)
Comment (laser) [reply proofs/sectorbuilder/interface.go]: done!

File: protocol/storage/prover.go (hunk 2 above)
Comment (laser) [reply protocol/storage/prover.go]: done!

File: protocol/storage/miner_test.go (hunk 1 above)
Comment (laser) [reply protocol/storage/miner_test.go]: done!

File: tools/iptb-plugins/filecoin/local/localfilecoin.go (hunk 4 above)
Comment (laser) [reply tools/iptb-plugins/filecoin/local/localfilecoin.go]: It is for debugging, yes. We've modified the default such that proofs logging is disabled by default.
